### PR TITLE
Expose conditional log prob method for enc-dec models.

### PR DESCRIPTION
**Pull request by MaximumEntropy**
Signed-off-by: MaximumEntropy <sandeep.subramanian.1@umontreal.ca>

# What does this PR do ?

Exposes a method `log_prob()` to get the conditional log prob `p(y|x)` for enc-dec models.

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -1110,18 +1110,7 @@ def dummy():
...
-        encoder_input=None,
-        tokenizer=None,
-        enc_output=None,
-        enc_output_attn_mask=None,
-        ignore_ids=[],
-        bos_id=None,  # If bos=None, will use tokenizer.bos_id unless explicitly set to something else.
-    ):
+    def _init_for_inference(self, tokens_enc, enc_output):
         # Check whether the DDP is initialized. This is needed when running inference outside of training loop.
         if parallel_state.is_unitialized():
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1115]: Is this piece of code used in other places? If so perhaps useful to put it in a utility method?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 2]
```diff
@@ -1259,6 +1265,107 @@ def dummy():
...
+        # TODO: Check if the user is trying to do gradient acc and maybe throw error
+        _reconfigure_microbatch_calculator(
+            rank=app_state.global_rank,
+            rampup_batch_size=None,
+            global_batch_size=global_batch_per_gpu * parallel_state.get_data_parallel_world_size(),
+            micro_batch_size=global_batch_per_gpu,  # Make sure that there is no "grad acc" while decoding.
+            data_parallel_size=parallel_state.get_data_parallel_world_size(),
+        )
+        # TODO: Figure out how to handle bos being either <bos> for NeMo-Megatron and <pad> for Huggingface/Google.
+        bos_id = tokenizer.bos_id if bos_id is None else bos_id
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1298]: Perhaps wrap this in a method that can be overridden to provide BOS token?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 3]
```diff
@@ -1259,6 +1265,107 @@ def dummy():
         )
         return predicted_tokens_dec, log_probs
 
+    def log_prob(
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1268]: Very useful method! Would it make sense to call `log_prob` within `decode`? They seem to have a lot of overlapping code regarding the logits computation

**Review by michalivne — COMMENTED** [review]
LGTM! See suggestions and minor comments
